test/e2e/e2e.go

```diff
@@ -1846,7 +1848,71 @@ var _ = ginkgo.Describe("e2e ingress traffic validation", func() {
+		// nodeport on the node with the backend actually receives traffic.
+		// In case of dual stack enabled cluster, we iterate over all the nodes ips and try to hit the
+		// endpoints from both each node's ips.
+		// TODO: Add Test that verifies SNAT=GWRouterIP of ingress traffic does not occur with externalTrafficPolicy=local
+		ginkgo.It("Should be allowed to node local endpoints by nodeport services with externalTrafficPolicy=local", func() {
+			serviceName := "nodeportsvclocal"
+			ginkgo.By("Creating the nodeport service")
+			// Put endpoints only on one node 
+			localNode := nodes.Items[0]
+			framework.Logf("The only node with an endpoint is: %s", localNode.Name)
+			localEndpintsSelector := map[string]string{"hostname": localNode.Name}
+			npSpec := nodePortServiceSpecFrom(serviceName, endpointHTTPPort, endpointUDPPort, clusterHTTPPort, clusterUDPPort, localEndpintsSelector,v1.ServiceExternalTrafficPolicyTypeLocal)
+			np, err := f.ClientSet.CoreV1().Services(f.Namespace.Name).Create(context.Background(), npSpec, metav1.CreateOptions{})
+			nodeTCPPort, nodeUDPPort := nodePortsFromService(np)
+			framework.ExpectNoError(err)
+			ginkgo.By("Waiting for the single endpoint to pop up on node " + localNode.Name)
+			err = framework.WaitForServiceEndpointsNum(f.ClientSet, f.Namespace.Name, serviceName, 1, time.Second, wait.ForeverTestTimeout)
+			framework.ExpectNoError(err, "failed to validate endpoints for service %s in namespace: %s", serviceName, f.Namespace.Name)
 
+			for _, protocol := range []string{"http", "udp"} {
+				for _, node := range nodes.Items {
+					for _, nodeAddress := range node.Status.Addresses {
+						// skipping hostnames
+						if !addressIsIP(nodeAddress) {
+							continue
+						}
+						responses := sets.NewString()
+						singleEndpointName := sets.NewString(localNode.Name + "-ep")
+						valid := false
+						nodePort := nodeTCPPort
+						if protocol == "udp" {
+							nodePort = nodeUDPPort
+						}
+						ginkgo.By("Hitting the nodeport on " + node.Name + " and trying to reach a single endpoint on Node " + localNode.Name + " with protocol "  + protocol )
+						for i := 0; i < maxTries; i++ {
+							epHostname := pokeEndpointHostname(clientContainerName, protocol, nodeAddress.Address, nodePort)
+							// There should only be a response from the node "localNode" which contains the nodeLocal Endpoint, otherwise
+							// traffic should "blackhole" 
+							if node.Name == localNode.Name {
+								framework.Logf("Contacting Local Endpoint on Node: %s response is %v", localNode.Name, epHostname)
+								responses.Insert(epHostname)
+							}
+							// If no node local endpoint poking endpoints should result in timeout
+							if epHostname == "Timeout" { 
+								framework.Logf("Contacting Local Endpoint on Node: %s response is %v", localNode.Name, epHostname)
+								responses.Insert(localNode.Name + "-ep")
+							}
+							if responses.Equal(singleEndpointName) {
```

Comment:
You may not want to break (maybe lowering the number of attempts) to avoid the luck (or not luck) of picking the right ep by chance.

Reply:
I added "no-SNAT" validation as well so it should be fine now, if ETP wasn't working both the hostname and packet src ip would be incorrect